packages/preferences/src/browser/ui/settings-ui-plus-fuzzy-search.ts

```diff
@@ -0,0 +1,69 @@
+/********************************************************************************
+ * Copyright (C) 2020 Ericsson and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import * as fuzzy from 'fuzzy';
+
+interface FilterOptions {
+    debug?: boolean;
+    childProp?: string;
+    minLength?: number;
+};
+
+const filterDefaults: Required<FilterOptions> = {
+    childProp: 'children',
+    debug: false,
+    minLength: 1,
+};
+
+function flattenTrees(trees: any[], childProp: string): any[] {
+    return trees.concat(...trees.map(node => flattenTrees(node[childProp] || [], childProp)));
+}
+
+function toFlatNameArray(list: any[], childProp: string): any[] {
+    return flattenTrees(list, childProp).map(node => node.name);
+}
+
+export function settingsUIPlusFuzzySearch(
```

Comment:
exporting function is anti-pattern in Theia codebase, please review DI coding guideline section

Theia is open framework and should allow subclassing and function overrides, if some code paradigms does not support it then they should not be used in Theia repo

Reply:
Does this also include functional React components used as abstractions for a widget? As an example, I'm talking about components such as SettingsNavbar.

Reply:
Can extenders rebind them and modify rendering? If there is another way like extenders can hook somewhere earlier and provide another rendering by not copy pasting to much, but composing such components it is fine, if not classes are preferable. At the end Theia is a framework and then everything is done in the same approach to extensibility it is better for extenders. Extenders have a freedom to use whatever they want.